[#165] Keycloak support added

filip-kopecky opened the pull request:
Added keycloak

filip-kopecky committed:
[#165] Keycloak

filip-kopecky committed:
spring active profile

filip-kopecky committed:
spaces

filip-kopecky committed:
Added profile to tests + removed isAuthenticated from endpoint

Added spring profile to relevant tests + removed isAuthenticated annotation where it wasn't required

filip-kopecky commented:
Added profiles to the relevant test classes + removed isAuthenticated from `/configuration` endpoint. The previously failed tests should be working now. @ledsoft

ledsoft commented on pom.xml:
```diff
@@ -454,6 +482,8 @@
                         </path>
                     </annotationProcessorPaths>
+                    <source>11</source>
+                    <target>11</target>
```
Replace with `java.version` from properties.

ledsoft commented on src/main/java/cz/cvut/kbss/termit/config/SecurityKeycloakConfig.java:
```diff
@@ -0,0 +1,84 @@
+        KeycloakAuthenticationProvider keycloakAuthenticationProvider
+                = keycloakAuthenticationProvider();
+        keycloakAuthenticationProvider.setGrantedAuthoritiesMapper(
+                new SimpleAuthorityMapper());
```
Replace with a custom authority mapper looking like this:
```java
public class HierarchicalRoleBasedAuthorityMapper implements GrantedAuthoritiesMapper {

    @Override
    public Collection<? extends GrantedAuthority> mapAuthorities(Collection<? extends GrantedAuthority> authorities) {
        return authorities.stream().map(a -> {
                              final UserRole role = UserRole.fromRoleName(a.getAuthority());
                              return role.getGranted();
                          }).flatMap(roles -> roles.stream().map(r -> new SimpleGrantedAuthority(r.getName())))
                          .collect(Collectors.toSet());
    }
}
```

ledsoft reviewed (changes requested):
I figured out the problem with `PreAuthorize` - TermIt internal roles work hierarchically, so `ROLE_FULL_USER` automatically includes `ROLE_RESTRICTED_USER`. Keycloak-based authorization does not work like this.
To fix it, create a custom authority mapper and pass it to `KeycloakAuthenticationProvider`. See the inline comment. After that, all the extra `PreAuthorize("isAuthenticated()")` can be removed and the PR will me much more compact.

ledsoft committed:
[#165] Implement custom authority mapper to support hierarchical roles.

Simplify CORS configuration - remove duplication.

ledsoft committed:
[#165] Remove unnecessary PreAuthorize statements.

Was fixed by custom authority mapper.

ledsoft committed:
[Ref] Replace hardcoded Java version with variable reference.

ledsoft committed:
[#165] Replace profile-based security configuration, use property condition.

Instead of having to provide profile, TermIt will determine whether to use Keycloak or internal security based on presence of Keycloak server config.

ledsoft committed:
[#165] Fix keycloak configuration, disable it by default.

ledsoft reviewed (approved):
LGTM
